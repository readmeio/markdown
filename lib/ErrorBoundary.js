import React from 'react';

export default class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return { hasError: true };
  }

  render() {
    if (this.state.hasError) {
      return null;
    }

    // eslint-disable-next-line react/prop-types
    return this.props.children;
  }
}
