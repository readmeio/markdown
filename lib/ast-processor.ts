import { remark } from 'remark';
import remarkMdx from 'remark-mdx';
import remarkFrontmatter from 'remark-frontmatter';
import remarkGfm from 'remark-gfm';

import transformers, { readmeComponentsTransformer, variablesTransformer } from '../processor/transform';
import rehypeSlug from 'rehype-slug';

export type MdastOpts = {
  components?: Record<string, string>;
};

export const remarkPlugins = [remarkFrontmatter, remarkGfm, ...transformers];
export const rehypePlugins = [rehypeSlug];

const astProcessor = (opts: MdastOpts = { components: {} }) =>
  remark()
    .use(remarkMdx)
    .use(remarkPlugins)
    .use(variablesTransformer, { asMdx: false })
    .use(readmeComponentsTransformer({ components: opts.components }));

export default astProcessor;
