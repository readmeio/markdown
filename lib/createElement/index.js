const React = require('react');

const { CodeTabs } = require('../../components/CodeTabs');

const createElement = (type, props, ...children) => {
  const rdmdType = type === 'div' && props?.className === 'code-tabs' ? CodeTabs : type;

  return React.createElement(rdmdType, props, ...children);
};

module.exports = createElement;
