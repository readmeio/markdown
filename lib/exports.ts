import type { MdxjsEsm } from 'mdast-util-mdx';

import { visit } from 'unist-util-visit';

import { isMDXEsm } from '../processor/utils';

import mdast from './mdast';

/* Example mdast structures to find first export name in a mdxjsEsm node:
There are three types of export declarations that we need to consider:
1. VARIABLE DECLARATION
      "type": "mdxjsEsm",
      "value": "export const Foo = () => <div>Hello world</div>\nexport const Bar = () => <div>hello darkness my old friend</div>",
      "data": {
        "estree": {
          "type": "Program",
          "body": [
            {
              "type": "ExportNamedDeclaration",
              "declaration": {
                "type": "VariableDeclaration",
                "declarations": [
                  {
                    "type": "VariableDeclarator",
                    "id": {
                      "type": "Identifier",
                      "name": "Foo" // --------> This is the export name
                    },
                    ...

2/3. FUNCTION DECLARATION & CLASS DECLARATION
      "estree": {
          "type": "Program",
          "body": [
            {
              "type": "ExportNamedDeclaration",
              "declaration": {
                "type": "ClassDeclaration" | "FunctionDeclaration",
                "id": {
                  "type": "Identifier",
                  "name": "Foo" // --------> This is the export name
                },
*/

const exports = (doc: string) => {
  const set = new Set<string>();

  visit(mdast(doc), isMDXEsm, (node: MdxjsEsm) => {
    // Once inside an mdxjsEsm node, we need to check for one or more declared exports within data.estree.body
    // This is because single newlines \n are not considered as a new block, so there may be more than one export statement in a single mdxjsEsm node
    const body = node.data?.estree.body;
    if (!body) return;

    body.forEach(child => {
      if (child.type === 'ExportNamedDeclaration') {
        // There are three types of ExportNamedDeclaration that we need to consider: VariableDeclaration, FunctionDeclaration, ClassDeclaration
        const declaration = child.declaration;

        // FunctionDeclaration and ClassDeclaration have the same structure
        if (declaration.type !== 'VariableDeclaration') {
          // Note: declaration.id.type is always 'Identifier' for FunctionDeclarations and ClassDeclarations
          set.add(declaration.id.name);
        } else {
          declaration.declarations.forEach(({ id }) => {
            if (id.type === 'Identifier') {
              set.add(id.name);
            }
          });
        }
      }
    });
  });

  return Array.from(set);
};

export default exports;
