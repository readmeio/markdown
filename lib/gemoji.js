const { nameToEmoji } = require('gemoji');

const owlmoji = ['owlbert-books', 'owlbert-mask', 'owlbert', 'owlbert-reading', 'owlbert-thinking'];

class Emoji {
  static kind = name => {
    if (name in nameToEmoji) return 'gemoji';
    else if (name.match(/^fa-/)) return 'fontawesome';
    else if (owlmoji.includes(name)) return 'owlmoji';
    return null;
  };

  static nameToEmoji = nameToEmoji;
}

module.exports = Emoji;
