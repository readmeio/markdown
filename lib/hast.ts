import astProcessor, { rehypePlugins, MdastOpts } from './ast-processor';
import remarkRehype from 'remark-rehype';
import { injectComponents } from '../processor/transform';
import { MdastComponents } from '../types';
import mdast from './mdast';

const hast = (text: string, opts: MdastOpts = {}) => {
  const components: MdastComponents = Object.entries(opts.components || {}).reduce((memo, [name, doc]) => {
    memo[name] = mdast(doc);
    return memo;
  }, {});

  const processor = astProcessor(opts).use(injectComponents({ components })).use(remarkRehype).use(rehypePlugins);

  return processor.runSync(processor.parse(text));
};
export default hast;
