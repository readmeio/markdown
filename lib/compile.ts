import { compile as mdxCompile, CompileOptions } from '@mdx-js/mdx';
import remarkFrontmatter from 'remark-frontmatter';
import remarkGfm from 'remark-gfm';

import { defaultTransforms, tailwindRootTransformer } from '../processor/transform';
import { rehypeToc } from '../processor/plugin/toc';
import MdxSyntaxError from '../errors/mdx-syntax-error';
import { rehypePlugins } from './ast-processor';
import { CustomComponents } from '../types';

export type CompileOpts = CompileOptions & {
  lazyImages?: boolean;
  safeMode?: boolean;
  components?: CustomComponents;
  copyButtons?: boolean;
};

const { codeTabsTransformer, ...transforms } = defaultTransforms;

const compile = async (text: string, { components = {}, copyButtons, ...opts }: CompileOpts = {}) => {
  try {
    const vfile = await mdxCompile(text, {
      outputFormat: 'function-body',
      providerImportSource: '#',
      remarkPlugins: [
        remarkFrontmatter,
        remarkGfm,
        ...Object.values(transforms),
        [codeTabsTransformer, { copyButtons }],
        [tailwindRootTransformer, { components }],
      ],
      rehypePlugins: [...rehypePlugins, [rehypeToc, { components }]],
      ...opts,
    });

    return String(vfile).replace('"use strict";', '"use strict";\nconst { React, user } = arguments[0].imports;\n');
  } catch (error) {
    throw error.line ? new MdxSyntaxError(error, text) : error;
  }
};

export default compile;
