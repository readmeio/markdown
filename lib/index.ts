import astProcessor, { MdastOpts, remarkPlugins } from './ast-processor';
import compile from './compile'
import mdx from './mdx';
import run from './run';

export type { MdastOpts };
export { astProcessor, compile, mdx, run, remarkPlugins }
