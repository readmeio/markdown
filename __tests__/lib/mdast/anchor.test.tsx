
import { mdast } from '../../../lib';

describe('convert anchor tag', () => {
    it('converts anchor tag to link node', () => {
    const mdx = `
<Anchor href="https://readme.com">ReadMe</Anchor>
    `;

    expect(mdast(mdx)).toMatchSnapshot();
  });
})
