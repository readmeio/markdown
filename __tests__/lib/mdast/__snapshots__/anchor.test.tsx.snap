// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convert anchor tag > converts anchor tag to link node 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 41,
                  "line": 2,
                  "offset": 41,
                },
                "start": {
                  "column": 35,
                  "line": 2,
                  "offset": 35,
                },
              },
              "type": "text",
              "value": "ReadMe",
            },
          ],
          "position": {
            "end": {
              "column": 50,
              "line": 2,
              "offset": 50,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "link",
          "url": "https://readme.com",
        },
      ],
      "position": {
        "end": {
          "column": 50,
          "line": 2,
          "offset": 50,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 3,
      "offset": 55,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
